### PR TITLE
Support known amount swap stubs

### PR DESCRIPTION
**Pull request by jodastephen**
Add support for fixed leg stubs
Rename StubCalculation to be Ibor specific
Extend FpML parsing

The design solution of a package-scoped `KnownAmountRateComputation` is not ideal, but it is a practical solution to implement this feature without a major rewrite of complex logic.

### CODE REVIEW

File: modules/product/src/main/java/com/opengamma/strata/product/swap/FixedRateStubCalculation.java [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+   */
+  public static FixedRateStubCalculation ofKnownAmount(CurrencyAmount knownAmount) {
+    ArgChecker.notNull(knownAmount, "knownAmount");
+    return new FixedRateStubCalculation(null, knownAmount);
+  }
+
+  //-------------------------------------------------------------------------
+  @ImmutableValidator
+  private void validate() {
+    if (fixedRate != null && knownAmount != null) {
```
Comment (cjkent) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/FixedRateStubCalculation.java]: Shouldn't this also validate that at least one is non-null?

Comment (jodastephen) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/FixedRateStubCalculation.java]: No, the object may have both as null, see the NONE constant.

File: modules/product/src/main/java/com/opengamma/strata/product/swap/IborRateStubCalculation.java [hunk 2]
```diff
@@ -64,16 +65,25 @@
...
-   * If the fixed rate is present, then {@code index} and {@code indexInterpolated} must not be present.
+   * If the fixed rate is present, then {@code knownAmount}, {@code index} and
+   * {@code indexInterpolated} must not be present.
    */
   @PropertyDefinition(get = "optional")
   private final Double fixedRate;
   /**
+   * The known amount to pay/receive for the stub.
+   * <p>
+   * If the known amount is present, then {@code fixedRate}, {@code index} and
```
Comment (cjkent) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/IborRateStubCalculation.java]: It doesn't look like the `validate` method has been updated to reflect this.

File: modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountNotionalPaymentPeriod.java [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+        throw new IllegalArgumentException(Messages.format(
+            "Payment currency {} must not equal notional currency {} when FX reset applies", payCcy, notionalCcy));
+      }
+      if (!fxResetObservation.getIndex().getCurrencyPair().contains(payCcy)) {
+        throw new IllegalArgumentException(Messages.format(
+            "Payment currency {} must be one of those in the FxReset index {}", payCcy, fxResetObservation.getIndex()));
+      }
+      if (!fxResetObservation.getIndex().getCurrencyPair().contains(notionalCcy)) {
+        throw new IllegalArgumentException(Messages.format(
+            "Notional currency {} must be one of those in the FxReset index {}", payCcy, fxResetObservation.getIndex()));
```
Comment (cjkent) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountNotionalPaymentPeriod.java]: This should use `notionalCcy`

File: modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountRateComputation.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+/**
+ * Defines a known amount of interest as a rate computation.
+ * <p>
+ * This is a special computation that represents a known amount instead of a rate.
+ * It is used to pass the known amount through the standard rate computation process.
+ * This computation is converted to a {@link KnownAmountNotionalPaymentPeriod} for pricing.
+ */
+@BeanDefinition(style = "light")
+final class KnownAmountRateComputation
+    implements RateComputation, ImmutableBean, Serializable {
```
Comment (cjkent) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountRateComputation.java]: `Serializable` but no version ID.

**Comment by cjkent** [comment]
That's everything

File: modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountRateComputation.java (hunk 4 above)
Comment (jodastephen) [inline modules/product/src/main/java/com/opengamma/strata/product/swap/KnownAmountRateComputation.java]: It is within the Joda-Bean generated block.

**Pull request merged**